### PR TITLE
app/TreeSearch: add fuzzy scoring for search results

### PR DESCRIPTION
**Pull request by mkovacs**
Improve tree search: sort results by their fuzzy match score.

It's live on staging3.sourcegraph.com

### CODE REVIEW

File: app/web_modules/sourcegraph/tree/TreeSearch.js [hunk 1]
```diff
@@ -331,14 +324,21 @@ class TreeSearch extends Container {
...
-			} else if (nextState.fileList && nextState.fileList.Files) {
-				nextState.fileResults = nextState.fileList.Files
-					.filter((f) => fuzzyMatchPath(nextState.query, f))
-					.map((f) => ({
-						name: f,
-						isDirectory: false,
-						url: urlToBlob(nextState.repo, nextState.rev, f),
-					}));
+			} else {
+				// NOTE(mate): moving this constant declaration inside the if-block seems to trigger
```
Comment (rothfels) [inline app/web_modules/sourcegraph/tree/TreeSearch.js]: Is the note necessary to leave in here? Sounds like you're just qualifying that this shouldn't be moved into the `if` block, but if what you're saying is true, then `flow` is going to complain at the next dev who decides to try, so there's no value in keeping the comment here.

**Comment by rothfels** [comment]
@mkovacs LGTM, let's get this merged.

**Comment by rothfels** [comment]
(Left one minor comment which you can consider prior to merging.)

File: app/web_modules/sourcegraph/tree/TreeSearch.js (hunk 1 above)
Comment (mkovacs) [inline app/web_modules/sourcegraph/tree/TreeSearch.js]: It took me about 10 minutes of head-scratching to figure out why Flow was complaining about `nextState.fileList.Files` being possibly undefined right after it's checked. I'm hoping that note will save somebody 10 minutes.

**Pull request merged**